apollo-router/src/spec/query/change.rs

```diff
@@ -366,30 +544,45 @@ impl<'a> Visitor for QueryHashVisitor<'a> {
         field_def: &ast::FieldDefinition,
         node: &executable::Field,
     ) -> Result<(), BoxError> {
+        "^VISIT_FIELD".hash(self);
+
         if !self.seen_introspection && (field_def.name == "__schema" || field_def.name == "__type")
```

Comment:
That fields only have a special meaning for the query root type.
```
{
   foo {
     __schema # <- not an introspection field here
   }
}
```
Technically it should be forbided to have such names:
https://spec.graphql.org/draft/#sec-Names.Reserved-Names

But I can't find code that implements this validation in `apollo-rs`
https://github.com/search?q=repo%3Aapollographql%2Fapollo-rs++%5C%22__%5C%22&type=code

Also, it's not present here:
https://github.com/apollographql/apollo-rs/blob/main/crates/apollo-compiler/src/validation/field.rs#L155

I will report it to apollo-rs but I also think we need to be sure those are at the query root to count as introspection.

Reply:
I am opening https://github.com/apollographql/apollo-rs/issues/903 to track it. This will not cause issue for this algorithm, because if `__schema` appears in the wrong place, then the only consequence is that the schema string will be entirely hashed in, so that would not add any collision risk

Reply:
Indeed we are missing a validation rule to forbid definitions with a `__` prefix: https://github.com/apollographql/apollo-rs/issues/743